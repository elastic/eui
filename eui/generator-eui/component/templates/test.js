import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test';

import { <%= componentName %> } from './<%= fileName %>';

describe('<%= componentName %>', () => {
  test('is rendered', () => {
    const component = render(
      <<%= componentName %> {...requiredProps} />
    );

    expect(component)
      .toMatchSnapshot();
  });
});
