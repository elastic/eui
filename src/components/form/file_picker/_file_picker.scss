/**
  * REMEMBER: --large modifiers must come last to override --compressed
  */

.euiFilePicker {
  @include euiFormControlSize($includeAlternates: true);
  position: relative;

  &.euiFilePicker--large {
    border-radius: $euiFormControlBorderRadius;
    overflow: hidden;
    height: auto;
  }

  &.euiFilePicker--large.euiFilePicker--compressed {
    border-radius: $euiFormControlCompressedBorderRadius;
  }
}

// The input is an invisible dropzone / button
.euiFilePicker__input {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  overflow: hidden;

  &:hover {
    cursor: pointer;
  }

  &:hover:disabled {
    cursor: not-allowed;
  }

  &:disabled {
    opacity: 0;
  }

  &:disabled ~ .euiFilePicker__prompt {
    color: $euiColorMediumShade;
  }
}

.euiFilePicker__icon {
  position: absolute;
  left: $euiSizeM;
  top: $euiSizeM;
  transition: transform $euiAnimSpeedFast $euiAnimSlightResistance;

  .euiFilePicker--compressed & {
    top: $euiSizeS;
    left: $euiSizeS;
  }

  .euiFilePicker--large & {
    position: static;
    margin-bottom: $euiSize;
  }
}

/**
  * 1. Don't block the user from dropping files onto the filepicker.
  * 2. Ensure space for import icon, loading spinner, and clear button (only if it has files)
  * 3. Delay focus gradient or else it will only partially transition while file chooser opens
  * 4. Static height so that it doesn't shift its surrounding contents around
  */
.euiFilePicker__prompt {
  @include euiFormControlDefaultShadow;
  @include euiFormControlWithIcon; /* 2 */
  height: $euiFormControlHeight;
  padding-top: $euiFormControlPadding;
  padding-right: $euiFormControlPadding;
  padding-bottom: $euiFormControlPadding;
  pointer-events: none; /* 1 */
  border-radius: $euiFormControlBorderRadius;

  // sass-lint:disable-block indentation
  transition:
    box-shadow $euiAnimSpeedFast ease-in,
    background-color $euiAnimSpeedFast ease-in,
    background-image $euiAnimSpeedFast ease-in,
    background-size $euiAnimSpeedFast ease-in $euiAnimSpeedFast; /* 3 */

  .euiFilePicker--compressed & {
    @include euiFormControlStyleCompressed($includeStates: false);
    @include euiFormControlWithIcon($compressed: true); /* 2 */
    height: $euiFormControlCompressedHeight;
    border-radius: $euiFormControlCompressedBorderRadius;
  }

  .euiFilePicker--large & {
    height: $euiFilePickerTallHeight; /* 4 */
    padding: 0 $euiSizeL;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }

  .euiFilePicker--large.euiFilePicker--compressed & {
    height: $euiFilePickerTallHeight - $euiSizeL; /* 4 */
  }

  .euiFilePicker-isInvalid & {
    @include euiFormControlInvalidStyle;
  }
}

.euiFilePicker__promptText {
  @include euiFontSizeS;
  @include euiTextTruncate;
  line-height: $euiSize;

  // make normal sized prompt look like placeholder
  .euiFilePicker:not(.euiFilePicker--large):not(.euiFilePicker-hasFiles) & {
    color: $euiColorMediumShade;
  }
}

.euiFilePicker__clearButton,
.euiFilePicker__loadingSpinner {
  position: absolute;
  right: $euiSizeM;
  top: $euiSizeM;

  .euiFilePicker--compressed & {
    top: $euiSizeS;
  }
}

/**
  * 1. Undo the pointer-events: none applied to the enclosing prompt.
  */
.euiFilePicker__clearButton {
  pointer-events: auto; /* 1 */

  .euiFilePicker:not(.euiFilePicker--large) & {
    @include euiFormControlLayoutClearIcon('.euiFilePicker__clearIcon');
  }

  .euiFilePicker--large & {
    position: relative;
    top: 0;
    right: 0;
  }
}

// Focus
.euiFilePicker__showDrop .euiFilePicker__prompt,
.euiFilePicker__input:focus + .euiFilePicker__prompt {
  @include euiFormControlFocusStyle;

  .euiFilePicker--compressed & {
    @include euiFormControlFocusStyle($borderOnly: true);
  }
}

// Disabled
.euiFilePicker__input:disabled + .euiFilePicker__prompt {
  @include euiFormControlDisabledStyle;
}

// Make space for the icons on the right
.euiFilePicker:not(.euiFilePicker--large) {
  &.euiFilePicker-isLoading .euiFilePicker__prompt,
  &.euiFilePicker-hasFiles .euiFilePicker__prompt {
    @include euiFormControlWithIcon(false, 'right'); /* 2 */
  }
}

// When the filepicker has files in it
.euiFilePicker-hasFiles .euiFilePicker__promptText {
  color: $euiTextColor;
}

// Large styles only
.euiFilePicker--large {
  // sass-lint:disable-block nesting-depth
  // Hover and focus
  .euiFilePicker__input:hover:not(:disabled),
  .euiFilePicker__input:focus {
    + .euiFilePicker__prompt {
      .euiFilePicker__promptText {
        // Adding the underline to the prompt text ensures the underline
        // color is the same as the current text color
        text-decoration: underline;
      }

      .euiFilePicker__icon {
        transform: scale(1.1);
      }
    }
  }

  // While dragging files over the dropzone
  &.euiFilePicker__showDrop .euiFilePicker__prompt {
    .euiFilePicker__promptText {
      text-decoration: underline;
    }

    .euiFilePicker__icon {
      transform: scale(1.1);
    }
  }

  &.euiFilePicker-hasFiles .euiFilePicker__promptText {
    font-weight: $euiFontWeightBold;
  }
}
