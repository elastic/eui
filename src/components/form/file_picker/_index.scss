@import 'variables';
@import 'file_picker';
