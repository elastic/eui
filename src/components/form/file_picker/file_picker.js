import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EuiValidatableControl } from '../validatable_control';
import { EuiButtonEmpty } from '../../button';
import { EuiProgress } from '../../progress';
import { EuiIcon } from '../../icon';
import { EuiI18n } from '../../i18n';
import { EuiLoadingSpinner } from '../../loading';

const displayToClassNameMap = {
  default: null,
  large: 'euiFilePicker--large',
};

export const DISPLAYS = Object.keys(displayToClassNameMap);

export class EuiFilePicker extends Component {
  static propTypes = {
    id: PropTypes.string,
    name: PropTypes.string,
    className: PropTypes.string,
    /**
     * The content that appears in the dropzone if no file is attached
     */
    initialPromptText: PropTypes.node,
    /**
     * Use as a callback to access the HTML FileList API
     */
    onChange: PropTypes.func,
    /**
     * Reduces the size to a typical (compressed) input
     */
    compressed: PropTypes.bool,
    /**
     * Size or type of display;
     * `default` for normal height, similar to other controls;
     * `large` for taller size
     */
    display: PropTypes.oneOf(DISPLAYS),
    fullWidth: PropTypes.bool,
    isInvalid: PropTypes.bool,
    isLoading: PropTypes.bool,
  };

  static defaultProps = {
    initialPromptText: 'Select or drag and drop a file',
    compressed: false,
    display: 'large',
  };

  constructor(props) {
    super(props);
    this.state = {
      promptText: null,
      isHoveringDrop: false,
    };
  }

  handleChange = filesSelected => {
    if (this.fileInput.files && this.fileInput.files.length > 1) {
      this.setState({
        promptText: `${this.fileInput.files.length} ${filesSelected}`,
      });
    } else if (this.fileInput.files.length === 0) {
      this.setState({ promptText: null });
    } else {
      this.setState({ promptText: this.fileInput.value.split('\\').pop() });
    }

    const { onChange } = this.props;

    if (onChange) {
      onChange(this.fileInput.files);
    }
  };

  removeFiles = e => {
    e.stopPropagation();
    e.preventDefault();
    this.fileInput.value = null;
    this.handleChange();
  };

  showDrop = () => {
    if (!this.props.disabled) {
      this.setState({ isHoveringDrop: true });
    }
  };

  hideDrop = () => {
    this.setState({ isHoveringDrop: false });
  };

  render() {
    return (
      <EuiI18n
        tokens={[
          'euiFilePicker.clearSelectedFiles',
          'euiFilePicker.filesSelected',
        ]}
        defaults={['Clear selected files', 'files selected']}>
        {([clearSelectedFiles, filesSelected]) => {
          const {
            id,
            name,
            initialPromptText,
            className,
            disabled,
            compressed,
            onChange,
            isInvalid,
            fullWidth,
            isLoading,
            display,
            ...rest
          } = this.props;

          const isOverridingInitialPrompt = this.state.promptText != null;

          const normalFormControl = display === 'default';

          const classes = classNames(
            'euiFilePicker',
            displayToClassNameMap[display],
            {
              euiFilePicker__showDrop: this.state.isHoveringDrop,
              'euiFilePicker--compressed': compressed,
              'euiFilePicker--fullWidth': fullWidth,
              'euiFilePicker-isInvalid': isInvalid,
              'euiFilePicker-isLoading': isLoading,
              'euiFilePicker-hasFiles': isOverridingInitialPrompt,
            },
            className
          );

          let clearButton;
          if (isLoading && normalFormControl) {
            // Override clear button with loading spinner if it is in loading state
            clearButton = (
              <EuiLoadingSpinner className="euiFilePicker__loadingSpinner" />
            );
          } else if (isOverridingInitialPrompt) {
            if (normalFormControl) {
              clearButton = (
                <button
                  type="button"
                  aria-label={clearSelectedFiles}
                  className="euiFilePicker__clearButton"
                  onClick={this.removeFiles}>
                  <EuiIcon className="euiFilePicker__clearIcon" type="cross" />
                </button>
              );
            } else {
              clearButton = (
                <EuiButtonEmpty
                  aria-label={clearSelectedFiles}
                  className="euiFilePicker__clearButton"
                  size="xs"
                  onClick={this.removeFiles}>
                  Remove
                </EuiButtonEmpty>
              );
            }
          } else {
            clearButton = null;
          }

          const loader = !normalFormControl && isLoading && (
            <EuiProgress size="xs" color="accent" position="absolute" />
          );

          return (
            <div className={classes}>
              <div className="euiFilePicker__wrap">
                <EuiValidatableControl isInvalid={isInvalid}>
                  <input
                    type="file"
                    id={id}
                    name={name}
                    className="euiFilePicker__input"
                    onChange={() => this.handleChange(filesSelected)}
                    ref={input => {
                      this.fileInput = input;
                    }}
                    onDragOver={this.showDrop}
                    onDragLeave={this.hideDrop}
                    onDrop={this.hideDrop}
                    disabled={disabled}
                    {...rest}
                  />
                </EuiValidatableControl>
                <div className="euiFilePicker__prompt">
                  <EuiIcon
                    className="euiFilePicker__icon"
                    type="importAction"
                    size={normalFormControl ? 'm' : 'l'}
                    aria-hidden="true"
                  />
                  <div className="euiFilePicker__promptText">
                    {this.state.promptText || initialPromptText}
                  </div>
                  {clearButton}
                  {loader}
                </div>
              </div>
            </div>
          );
        }}
      </EuiI18n>
    );
  }
}
