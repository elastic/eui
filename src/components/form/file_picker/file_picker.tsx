/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { Component, InputHTMLAttributes, ReactNode } from 'react';
import classNames from 'classnames';

import { CommonProps, keysOf } from '../../common';

import { EuiValidatableControl } from '../validatable_control';
import { EuiButtonEmpty } from '../../button';
import { EuiProgress } from '../../progress';
import { EuiIcon } from '../../icon';
import { EuiI18n } from '../../i18n';
import { EuiLoadingSpinner } from '../../loading';
import { htmlIdGenerator } from '../../../services/accessibility';

const displayToClassNameMap = {
  default: null,
  large: 'euiFilePicker--large',
};

export const DISPLAYS = keysOf(displayToClassNameMap);

export type EuiFilePickerDisplay = keyof typeof displayToClassNameMap;

export interface EuiFilePickerProps
  extends CommonProps,
    Omit<InputHTMLAttributes<HTMLInputElement>, 'onChange'> {
  id?: string;
  name?: string;
  className?: string;
  /**
   * The content that appears in the dropzone if no file is attached
   */
  initialPromptText?: ReactNode;
  /**
   * Use as a callback to access the HTML FileList API
   */
  onChange?: (files: FileList | null) => void;
  /**
   * Reduces the size to a typical (compressed) input
   */
  compressed?: boolean;
  /**
   * Size or type of display;
   * `default` for normal height, similar to other controls;
   * `large` for taller size
   */
  display?: EuiFilePickerDisplay;
  fullWidth?: boolean;
  isInvalid?: boolean;
  isLoading?: boolean;
  disabled?: boolean;
}

export class EuiFilePicker extends Component<EuiFilePickerProps> {
  static defaultProps = {
    initialPromptText: 'Select or drag and drop a file',
    compressed: false,
    display: 'large',
  };

  state = {
    promptText: null,
    isHoveringDrop: false,
  };

  fileInput: HTMLInputElement | null = null;

  handleChange = (filesSelected?: string | null) => {
    if (!this.fileInput) return;

    if (this.fileInput.files && this.fileInput.files.length > 1) {
      this.setState({
        promptText: `${this.fileInput.files.length} ${filesSelected}`,
      });
    } else if (this.fileInput.files && this.fileInput.files.length === 0) {
      this.setState({ promptText: null });
    } else {
      this.setState({ promptText: this.fileInput.value.split('\\').pop() });
    }

    const { onChange } = this.props;

    if (onChange) {
      onChange(this.fileInput.files);
    }
  };

  removeFiles = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();
    e.preventDefault();

    if (!this.fileInput) return;

    this.fileInput.value = '';
    this.handleChange(null);
  };

  showDrop = () => {
    if (!this.props.disabled) {
      this.setState({ isHoveringDrop: true });
    }
  };

  hideDrop = () => {
    this.setState({ isHoveringDrop: false });
  };

  render() {
    return (
      <EuiI18n
        tokens={[
          'euiFilePicker.clearSelectedFiles',
          'euiFilePicker.filesSelected',
        ]}
        defaults={['Clear selected files', 'files selected']}>
        {([clearSelectedFiles, filesSelected]: string[]) => {
          const {
            id,
            name,
            initialPromptText,
            className,
            disabled,
            compressed,
            onChange,
            isInvalid,
            fullWidth,
            isLoading,
            display,
            ...rest
          } = this.props;

          let promptId: string = htmlIdGenerator()();

          if (id) {
            promptId = `${id}-filePicker__prompt`;
          }

          const isOverridingInitialPrompt = this.state.promptText != null;

          const normalFormControl = display === 'default';

          const classes = classNames(
            'euiFilePicker',
            displayToClassNameMap[display!],
            {
              euiFilePicker__showDrop: this.state.isHoveringDrop,
              'euiFilePicker--compressed': compressed,
              'euiFilePicker--fullWidth': fullWidth,
              'euiFilePicker-isInvalid': isInvalid,
              'euiFilePicker-isLoading': isLoading,
              'euiFilePicker-hasFiles': isOverridingInitialPrompt,
            },
            className
          );

          let clearButton;
          if (isLoading && normalFormControl) {
            // Override clear button with loading spinner if it is in loading state
            clearButton = (
              <EuiLoadingSpinner className="euiFilePicker__loadingSpinner" />
            );
          } else if (isOverridingInitialPrompt) {
            if (normalFormControl) {
              clearButton = (
                <button
                  type="button"
                  aria-label={clearSelectedFiles}
                  className="euiFilePicker__clearButton"
                  onClick={this.removeFiles}>
                  <EuiIcon className="euiFilePicker__clearIcon" type="cross" />
                </button>
              );
            } else {
              clearButton = (
                <EuiButtonEmpty
                  aria-label={clearSelectedFiles}
                  className="euiFilePicker__clearButton"
                  size="xs"
                  onClick={this.removeFiles}>
                  <EuiI18n
                    token="euiFilePicker.removeSelected"
                    default="Remove"
                  />
                </EuiButtonEmpty>
              );
            }
          } else {
            clearButton = null;
          }

          const loader = !normalFormControl && isLoading && (
            <EuiProgress size="xs" color="accent" position="absolute" />
          );

          return (
            <div className={classes}>
              <div className="euiFilePicker__wrap">
                <EuiValidatableControl isInvalid={isInvalid}>
                  <input
                    type="file"
                    id={id}
                    name={name}
                    className="euiFilePicker__input"
                    onChange={() => this.handleChange(filesSelected)}
                    ref={(input) => {
                      this.fileInput = input;
                    }}
                    onDragOver={this.showDrop}
                    onDragLeave={this.hideDrop}
                    onDrop={this.hideDrop}
                    disabled={disabled}
                    aria-describedby={promptId}
                    {...rest}
                  />
                </EuiValidatableControl>
                <div className="euiFilePicker__prompt" id={promptId}>
                  <EuiIcon
                    className="euiFilePicker__icon"
                    type="importAction"
                    size={normalFormControl ? 'm' : 'l'}
                    aria-hidden="true"
                  />
                  <div className="euiFilePicker__promptText">
                    {this.state.promptText || initialPromptText}
                  </div>
                  {clearButton}
                  {loader}
                </div>
              </div>
            </div>
          );
        }}
      </EuiI18n>
    );
  }
}
