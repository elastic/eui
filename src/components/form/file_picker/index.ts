export { EuiFilePicker, EuiFilePickerProps } from './file_picker';
