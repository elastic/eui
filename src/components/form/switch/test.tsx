import React from 'react';

import { EuiSwitch } from './switch';

export const Test = ({ checked }: { checked: boolean }) => (
  <EuiSwitch
    checked={checked}
    label="Test"
    onChange={e => console.log(e.target.checked)}
  />
);
