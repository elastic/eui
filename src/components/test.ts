export function add(x: number, y: number) {
  return x + y;
}

add(5, 4);
