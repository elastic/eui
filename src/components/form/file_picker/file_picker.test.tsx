import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test';

import { EuiFilePicker } from './file_picker';

// Mock the htmlIdGenerator to generate predictable ids for snapshot tests
jest.mock('../../../services/accessibility/html_id_generator', () => ({
  htmlIdGenerator: () => () => 'htmlId',
}));

describe('EuiFilePicker', () => {
  test('is rendered', () => {
    const component = render(<EuiFilePicker {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });
});
