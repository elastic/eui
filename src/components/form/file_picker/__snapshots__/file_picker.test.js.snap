// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiFilePicker is rendered 1`] = `
<div
  class="euiFilePicker euiFilePicker--large testClass1 testClass2"
>
  <div
    class="euiFilePicker__wrap"
  >
    <input
      aria-label="aria-label"
      class="euiFilePicker__input"
      data-test-subj="test subject string"
      type="file"
    />
    <div
      class="euiFilePicker__prompt"
    >
      <svg
        aria-hidden="true"
        class="euiIcon euiIcon--large euiIcon-isLoading euiFilePicker__icon"
        focusable="false"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      />
      <div
        class="euiFilePicker__promptText"
      >
        Select or drag and drop a file
      </div>
    </div>
  </div>
</div>
`;
